### PR TITLE
Fix/binding indexes

### PR DESCRIPTION
**Pull request by pixelspark**
This PR clarifies the 'optimisation' code a bit and also fixes an issue where incorrect binding indexes are applied to inputs when there are more than 4 (for outputs the binding index was defined as `binding_counter % 4` but it was just `binding_counter` for inputs).

### CODE REVIEW

File: src/optimisation.rs [hunk 1]
```diff
@@ -112,17 +112,19 @@ lazy_static! {
     };
 }
 
+const MAX_BINDINGS_PER_GROUP: usize = 4;
```
Comment (haixuanTao) [inline src/optimisation.rs:115]: Can we use: https://docs.rs/wgpu/latest/wgpu/struct.Limits.html wgpu limits constant with the defaults level, using `lazy_static`?

Comment (pixelspark) [reply src/optimisation.rs:115]: That seems like a good idea. However we should probably get the actual limits from the adapter/device at runtime (see [here](https://docs.rs/wgpu/latest/wgpu/struct.Adapter.html#method.limits)).

Comment (pixelspark) [reply src/optimisation.rs:115]: On second thought we should probably just stick to defining our own limit as 4: our shaders are all written with the maximum of 4 in mind, so even if wgpu would increase the limits we cannot simply assume it without changing the shader source (i.e. group/binding id's would have to be changed).

We could of course make this dynamic by getting the limit at runtime, adding that limit as a template variable, and dynamically determining group/binding id's. Unless there is a great performance benefit in having larger or less groups I am not sure that is worth the trouble.

Comment (haixuanTao) [reply src/optimisation.rs:115]: Well, the benefit of large bind group is not very clear from my point of you at the moment.

Well, agreed that the shaders are now written for bind group of size 4 and making it dynamic might break things. We can leave it as is for now, and change it later if we want to do optimisation on bind groups.

I can merge then, if it is ok for you.

Thanks for the good discussion.

**Comment by pixelspark** [comment]
OK, should be ready to merge now without conflicts! 👍

**Pull request merged**